### PR TITLE
Feat/add vesting backend and frontend

### PR DESCRIPTION
**Pull request by djeck1432**
1. Added new routing
2. Added new logic for `getEvents` 
3. Added new `.env.example` file

### CODE REVIEW

File: backend/.env.example [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# Env variables
+NODE_URL=#
+# If network == `mainnet` it will use StarknetChainId.SN_MAIN
+# otherwise it will use by default StarknetChainId.SN_SEPOLIA
+NETWORK=#
```
Comment (tensojka) [inline backend/.env.example:5]: For the example, I would put in a public NODE_URL such as https://free-rpc.nethermind.io/sepolia-juno

File: backend/src/starknet.ts [hunk 2]
```diff
@@ -1,24 +1,78 @@
...
 
-  const starknetIdNavigator = new StarknetIdNavigator(
+const starknetIdNavigator = new StarknetIdNavigator(
     provider,
-    constants.StarknetChainId.SN_MAIN
-  );
+    CHAIN_ID,
+);
+
+export const getVestingEvents = async (address: string) => {
```
Comment (tensojka) [inline backend/src/starknet.ts:22]: This is returning events, but what we would like to receive on the FE is the list of current active vests – why not do the 'compaction' of this on the backend?

**Review by tensojka — COMMENTED** [review]
Did you test it? How's the latency?

File: backend/.env.example (hunk 1 above)
Comment (djeck1432) [reply backend/.env.example:5]: 1. I tested it with mainnet for `all events` - ok 
2. I tested it with sepolia for `specific address` - ok

File: backend/src/starknet.ts (hunk 2 above)
Comment (djeck1432) [reply backend/src/starknet.ts:22]: It's returning `compaction` of event data,
so we will display, claimed, available for claim and future claim
that is why I made this return structure list of dicts:
```
amount: amount,
timestamp: timestamp,
is_claimable: true
```

**Review by tensojka — APPROVED** [review]

**Review by tensojka — APPROVED** [review]

File: backend/src/starknet.ts [hunk 3]
```diff
@@ -1,24 +1,126 @@
...
+
+      // Update `fromBlock` to the next block after the last fetched block
+      const lastEventBlock = events.events[events.events.length - 1].block_number;
+      fromBlock = lastEventBlock + 1; // Move to the next block
+
+      await new Promise(resolve => setTimeout(resolve, 100));
+    }
+
+    // Cache the result for future requests
+    cache.set(cacheKey, allEvents);
```
Comment (tensojka) [inline backend/src/starknet.ts:107]: Is there a TTL in the cache? What if the backend has been running for a few months and then new vesting events have been added? Stale data will be served until a BE restart.

File: frontend/src/components/VestingTable.tsx [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+            const response = await axios.get(`${BASE_API_URL}vesting-events`, {
+                params: {
+                    contract: CONTRACT_ADDR,
+                    address: address,
+                }
+            });
+            return response.data;
+        },
+        enabled: !!address,  // Only fetch if the address is available
+        retry: false,        // Disable retries for this query
```
Comment (tensojka) [inline frontend/src/components/VestingTable.tsx:36]: Great idea to use React Query 👍

File: frontend/src/components/VestingTable.tsx [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    const { address } = useAccount();
+    const [currentPage, setCurrentPage] = useState(1); // Track the current page
+
+    const { data: events = [], isLoading, error } = useQuery({
+        queryKey: ['vesting-events', address],
+        queryFn: async () => {
+            if (!address) {
+                return [];
+            }
+            const response = await axios.get(`${BASE_API_URL}vesting-events`, {
```
Comment (tensojka) [inline frontend/src/components/VestingTable.tsx:27]: see `apiService.ts` line 10:

`${BASE_API_URL}/submit`

Unify it in both places such that one BASE_API_URL can work for both.

File: frontend/src/components/VestingTable.tsx [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    if (error) {
+        return <p>{(error as Error).message || 'Failed to load vesting events.'}</p>;
+    }
+
+    return (
+        <div>
+            {/* Title */}
+            <div className="flex w-full flex-grow pb-4 text-2xl font-bold">Vesting Event</div>
+
+            {/* Table */}
```
Comment (tensojka) [inline frontend/src/components/VestingTable.tsx:69]: In the review for transfer status, I pointed out that that was the third table-like entity on the page that looks different. This is the fourth table-like entity on the page now – and yet it looks different from all the others. Please make it look consistent with at least one of the others, preferably the Your Stakes I think. (But your call.)

**Review by tensojka — CHANGES_REQUESTED** [review]

**Review by tensojka — COMMENTED** [review]
I have rewritten the event fetching and cache structure in the backend into a more maintainable form as I found it hard to communicate what changes were needed (see the do..while commit which did not lead to code that would be any more maintainable).

The cache structure was also improved.

File: backend/src/starknet.ts [hunk 7]
```diff
@@ -63,18 +63,30 @@ export const getVestingEvents = async (contract: string, address: string): Promi
...
+    // Filter out isVestingMilestone if it has a corresponding isVested event with the same claimable_at timestamp
+    const filteredEvents = events.filter((event, index, self) => {
+      if (event.is_claimable) {
+        const matchingVested = self.find(e => e.is_claimed && e.claimable_at === event.claimable_at);
+        return !matchingVested;
+      }
+      return true;
+    });
+
+    return filteredEvents;
```
Comment (tensojka) [inline backend/src/starknet.ts:88]:
```
const vestedClaimableAts = new Set(
    events.filter(e => e.is_claimed).map(e => e.claimable_at)
);

const filteredEvents = events.filter(event => 
    !event.is_claimable || !vestedClaimableAts.has(event.claimable_at)
);

return filteredEvents;
```

File: backend/src/starknet.ts [hunk 8]
```diff
@@ -1,24 +1,148 @@
...
+      const newEvents = events.events.reduce((acc: any[], event: any) => {
+        try {
+          const grantee = event.data[0]; // Grantee (index 0)
+          const timestamp = Number(BigInt(event.data[1]));  // Timestamp (index 1)
+          let amount = Number(BigInt(event.data[2]));  // Amount as BigInt (index 2)
+
+          // Apply scaling if amount > 0
+          if (amount > 0) {
+            amount = amount / (10 ** 18);
+          }
```
[the remark this answers is not in the document]
Comment (djeck1432) [reply backend/src/starknet.ts]: resolved in dm

**Review by tensojka — APPROVED** [review]

**Pull request merged**